SimTKcommon/include/SimTKcommon/internal/Value.h

```diff
@@ -32,72 +32,204 @@
 #include <limits>
 #include <typeinfo>
 #include <sstream>
+#include <utility>
 
 namespace SimTK {
+  // Forward declaration.
+  template<typename T> class Value;
 
-/**
- * Abstract base class representing an arbitrary value of self-describing type.
- */
+//==============================================================================
+//                             ABSTRACT VALUE
+//==============================================================================
+/** Abstract base class representing an arbitrary value of unknown type.
+This provides an ability to manipulate "values" abstractly without knowing the
+specific type of that value. This is useful, for example, for discrete state
+variables and Measures, where much of what we need to do with them is 
+independent of their types. **/
 class AbstractValue {
 public:
-    AbstractValue() { }
-    virtual ~AbstractValue() { }
-
-    virtual String      getTypeName() const = 0;    
-    virtual String        getValueAsString() const = 0;
-    virtual bool        isCompatible(const AbstractValue&) const = 0;
-    virtual void compatibleAssign(const AbstractValue&) = 0;
-        
-    AbstractValue& operator=(const AbstractValue& v) { compatibleAssign(v); return *this; }
-    
+    /** Create a deep copy of this object. **/
     virtual AbstractValue* clone() const = 0;
+
+    /** Return a human-readable form of the object type that is stored in the
+    concrete derived class underlying this %AbstractValue. **/
+    virtual String getTypeName() const = 0;  
+
+    /** Return a human-readable representation of the value stored in this
+    %AbstractValue object. **/
+    virtual String getValueAsString() const = 0;
+
+    /** Check whether the `other` object contains a value that is assignment-
+    compatible with this one. If so you can perform the assignment with the
+    assignment operator or with compatibleAssign(). **/
+    virtual bool isCompatible(const AbstractValue& other) const = 0;
+
+    /** If the `source` contains a compatible value, assign a copy of
+    that value into this object. Otherwise an exception is thrown. **/
+    virtual void compatibleAssign(const AbstractValue& source) = 0;
+    
+    /** Invokes the compatibleAssign() method which will perform the assignment
+    if the source object is compatible, or throw an exception otherwise. **/
+    AbstractValue& operator=(const AbstractValue& v) 
+    {   compatibleAssign(v); return *this; }
+
+    /** Retrieve the original (type-erased)*thing* as read-only. The template
+    argument must be exactly the non-reference type of the stored *thing*. **/
```

Comment:
If you mean to refer to the literal variable name `thing` in a doxygen comment use code font rather than italics, that is use back single quotes (\`thing\`) rather than italics (\*thing\*). If you just meant this generically then italics is fine, but might be unclear because the `AbstractValue` user won't necessarily know that `Value<T>` calls its internal T a "thing". Maybe call it a "type T object"?